### PR TITLE
[expo-updates] Move runtime version calculation ahead of project mutations

### PR DESCRIPTION
**Pull request by wschurman**
# Why

Generic projects (including those that do continuous native generation, CNG) are having some issues with fingerprint when built on EAS build. This is because EAS build does mutations of the native files just before build, thus generating a different logical runtime from the fingerprint perspective. (https://github.com/expo/eas-build/blob/main/packages/build-tools/src/android/expoUpdates.ts#L55 for example). https://github.com/expo/expo/pull/27576 was one possible solution, but this PR (in combination with a eas-build PR) is the better solution.

# What to review

- Is there a precedent for passing data between build phases?

# How

This PR (in combination with https://github.com/expo/expo/pull/27597) calculates the fingerprint runtime version ahead of project mutation so that a consistent fingerprint is calculated with the local version.

# Test Plan

Android:
1. Create project using latest canaries
2. Make changes to `eas-build` locally to set this environment variable during a build step before project modification.
3. Run build locally using `eas-build` local instructions.
4. See the following log: `[RUN_GRADLEW] Using fingerprint from EXPO_UPDATES_FINGERPRINT_OVERRIDE: a8715b7f12f5736af20ed1f66daa693d37c43585`, indicating the the override is respected and used
5. Check built aab for overridden fingerprint

iOS: Same, but look for the logs.

Test with a normal SDK 50 project with expo-updates to ensure no regressions.

### CODE REVIEW

**Review by quinlanj — APPROVED** [review]

**Comment by brentvatne** [comment]
i'm ok with passing this value around between steps, but i defer to @sjchmiela and @szdziedzic on the specifics of the approach for doing so.

we'll want to figure out how we express this cleanly as a "custom build" also -- i imagine it'll just be a separate runtime version calculation step, but one tricky thing here is that people might forget to add that step and only use the eas update configuration step.

**Review by szdziedzic — APPROVED** [review]
It looks reasonable to me and I think that if my comments don't concern you we can go ahead and merge it.

I think that we will need to update the `BuildPhase`s on the website to parse the logs correctly. When looking at it I noticed that we unnecessarily maintain the same enum in 2 places so I opened a PR to improve it: https://github.com/expo/universe/pull/14719. Once this PR lands and new libs versions are published I can bump the `@expo/eas-build-job` version in my PR. 

> Is there a precedent for passing data between build phases?

We already do so for iOS credentials so it shouldn't be a problem in my opinion.

> we'll want to figure out how we express this cleanly as a "custom build" also -- i imagine it'll just be a separate runtime version calculation step, but one tricky thing here is that people might forget to add that step and only use the eas update configuration step.

Yeah, I also think that we should turn it into a separate runtime version calculation step. So we also need to create a new custom build function and update our examples + docs. I can do it.

So the deployment order should be:
1. This PR -> `main`
2. PR with custom build function -> `main` (@szdziedzic) 
2. Release the new versions of libs (@szdziedzic)
3. Update `@expo/eas-build-version` in https://github.com/expo/universe/pull/14719 and deploy to prod (@szdziedzic) 
4. Update the libs in Turtle v2 and deploy to prod (@szdziedzic)

**Comment by szdziedzic** [comment]
I created the https://github.com/expo/eas-build/pull/362 so the same thing is supported in custom builds as well

File: packages/build-tools/src/builders/ios.ts
Comment (wschurman) [reply packages/build-tools/src/builders/ios.ts:85]: This is an excellent point. From what I can tell this isn't used anymore. If it is (or we need to keep it around for old projects that do use it, this seems likely) and it produces a different fingerprint, that is most definitely an issue.

Originally we had discussed building this as a job field, where the runtime version from the job field would be set to override the build fingerprint. So in this case the fingerprint would be calculated on the local machine and sent to EAS Build. The downside of this is that it's possible for EAS Build to change the logical runtime via things like prebuild scripts. So we'd be providing the user with a potential footgun. Thus, I've been trying to keep this as true as possible to just being the fingerprint of whatever is being built at compile time. Of course this PR breaks that, so maybe it's okay to break that a little further.

File: packages/build-tools/src/builders/android.ts
Comment (wschurman) [reply packages/build-tools/src/builders/android.ts:77]: The runtimeVersion passed in is currently only used for showing a warning to the user if the computed fingerprint differs from what the local machine thought it was. 

We need to calculate and override the fingerprint even if the user passed in a runtimeVersion. That being said, from my response to your comment above, we could start trusting and using this field as the override fingerprint rather than re-calculating it.

This is tough.

**Comment by wschurman** [comment]
Just tested with the most recent canary and it looks like the `APPLE_TEAM_ID` might not be used? Grep'ing for `APPLE_TEAM_ID` in the expo/expo repo (prebuild code) doesn't turn up anything. I'll land this as-is and we can always tweak later if necessary.

@szdziedzic - thanks for the detailed review! I'll land this and then review your PR addressing custom builds.

**Pull request merged**
